File: src/main/java/catwalks/Const.java
package catwalks;

import catwalks.block.property.UPropertyEnum;
import catwalks.block.property.UPropertyObject;
import catwalks.part.data.CatwalkRenderData;
import net.minecraft.block.properties.PropertyEnum;
import net.minecraft.launchwrapper.Launch;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.math.Vec3d;

import java.util.Random;

import static catwalks.util.meta.MetaStorage.bits;

public class Const {
	
	{ /* Centeralizing constants */ }
	
	public static int MATERIAL_BITS = bits(128);
	
	{ /* DRYing constants */ }
	
	public static Random RAND = new Random();
	
	public static EnumFacing[] HORIZONTALS_FROM_NORTH = new EnumFacing[] { EnumFacing.NORTH, EnumFacing.EAST, EnumFacing.SOUTH, EnumFacing.WEST };
	
	public static boolean developmentEnvironment = (Boolean)Launch.blackboard.get("fml.deobfuscatedEnvironment");
	public static final String MODID = CatwalksMod.MODID;
	
	public static final Vec3d VEC_CENTER = new Vec3d(0.5, 0.5, 0.5);
	public static final Vec3d VEC_ANTICENTER = new Vec3d(-0.5, -0.5, -0.5);
	
	public static final ResourceLocation ENTITY_DATA_CAPABILITY_LOC = new ResourceLocation(MODID, "entityData");
//
//	@CapabilityInject(ICWEntityData.class)
//	public static Capability<ICWEntityData> CW_ENTITY_DATA_CAPABILITY = null;
	
	{	/* blockstate properties*/	}
	
	public static final UPropertyEnum<EnumFacing> FACING = UPropertyEnum.create("facing", EnumFacing.class);
	public static final PropertyEnum<EnumCatwalkMaterial> MATERIAL = PropertyEnum.create("material", EnumCatwalkMaterial.class);
	
	{ /* render-only properties (unlisted properties) */ }
	
	public static final String COMMAND_OPTIONS = "options";
	
	public static final UPropertyObject<CatwalkRenderData> CATWALK_RENDER_DATA = new UPropertyObject<CatwalkRenderData>("renderdata", CatwalkRenderData.class);
	
	public static class NODE {
		private static int i = -1;
		public static final int PITCH  = i--;
		public static final int YAW  = i--;
		public static final int CONNECT_POINT = i--;
		public static final int OPEN_GUI = i--;
	};
	
	public static class GUI {
		public static int NODE_MANIPULATOR = -1;
	}
	
	public static ResourceLocation location(String path) {
		return new ResourceLocation(MODID, path);
	}
}

File: src/main/java/catwalks/Const.kt
package catwalks

import catwalks.block.property.UPropertyEnum
import catwalks.block.property.UPropertyObject
import catwalks.part.data.CatwalkRenderData
import catwalks.part.data.ScaffoldRenderData
import net.minecraft.block.properties.PropertyEnum
import net.minecraft.launchwrapper.Launch
import net.minecraft.util.EnumFacing
import net.minecraft.util.ResourceLocation
import net.minecraft.util.math.Vec3d
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import java.util.*

object Const {

    init { /* Centeralizing constants */
    }

    var MATERIAL_BITS = 128.combination_bits()

    init { /* DRYing constants */
    }

    val log: Logger = LogManager.getLogger(CatwalksMod.MODID)

    val RAND = Random()

    val HORIZONTALS_FROM_NORTH = arrayOf(EnumFacing.NORTH, EnumFacing.EAST, EnumFacing.SOUTH, EnumFacing.WEST)

    var developmentEnvironment = Launch.blackboard["fml.deobfuscatedEnvironment"] as Boolean
    val MODID = CatwalksMod.MODID

    val VEC_CENTER = Vec3d(0.5, 0.5, 0.5)
    val VEC_ANTICENTER = Vec3d(-0.5, -0.5, -0.5)

    val ENTITY_DATA_CAPABILITY_LOC = ResourceLocation(MODID, "entityData")
    //
    //	@CapabilityInject(ICWEntityData.class)
    //	public static Capability<ICWEntityData> CW_ENTITY_DATA_CAPABILITY = null;

    init {    /* blockstate properties*/
    }

    val FACING = UPropertyEnum.create<EnumFacing>("facing", EnumFacing::class.java)
    val MATERIAL = PropertyEnum.create<EnumCatwalkMaterial>("material", EnumCatwalkMaterial::class.java) {
        it?.STATUS?.shouldRegister ?: false
    }

    init { /* render-only properties (unlisted properties) */
    }

    val CATWALK_RENDER_DATA = UPropertyObject<CatwalkRenderData>("renderdata", CatwalkRenderData::class.java)
    val SCAFFOLD_RENDER_DATA = UPropertyObject<ScaffoldRenderData>("renderdata", ScaffoldRenderData::class.java)

    fun location(path: String): ResourceLocation {
        return ResourceLocation(MODID, path)
    }
}
